src/plugins/ds_mongo.c

```diff
@@ -0,0 +1,3126 @@
+}
+
+/* get name of the current process user */
+static sr_error_info_t *
+srpds_uid2usr(uid_t uid, char **username)
+{
+    sr_error_info_t *err = NULL;
+    int r;
+    struct passwd pwd, *pwd_p;
+    char *buf = NULL, *mem;
+    ssize_t buflen = 0;
+
+    assert(username);
+
+    do {
+        if (!buflen) {
+            // learn suitable buffer size
+            buflen = sysconf(_SC_GETPW_R_SIZE_MAX);
+            if (buflen == -1) {
+                buflen = 2048;
+            }
+        } else {
+            // enlarge buffer
+            buflen += 2048;
+        }
+
+        // allocate some buffer
+        mem = realloc(buf, buflen);
+        if (!mem) {
+            WRAPERR(&err, SR_ERR_NO_MEMORY, "realloc()", "")
+            goto cleanup;
+        }
+        buf = mem;
+
+        // UID -> user
+        r = getpwuid_r(uid, &pwd, buf, buflen, &pwd_p);
+    }
+    while (r == ERANGE);
+
+    if (r) {
+        WRAPERR(&err, SR_ERR_INTERNAL, "Retrieving UID passwd entry", strerror(r))
+        goto cleanup;
+    }
+    else if (!pwd_p) {
+        WRAPERR(&err, SR_ERR_NOT_FOUND, "Retrieving UID passwd entry (No such UID)", "")
+        goto cleanup;
+    }
+
+    *username = strdup(pwd.pw_name);
+    if (!*username) {
+        WRAPERR(&err, SR_ERR_NO_MEMORY, "strdup()", strerror(errno))
+        goto cleanup;
+    }
+
+cleanup:
+    free(buf);
+    return err;
+}
+
+/* get name of the current process group */
+static sr_error_info_t *
+srpds_gid2grp(gid_t gid, char **group)
+{
+    sr_error_info_t *err = NULL;
+    int r;
+    struct group grp, *grp_p;
+    char *buf = NULL, *mem;
+    ssize_t buflen = 0;
+
+    assert(group);
+
+    do {
+        if (!buflen) {
+            // learn suitable buffer size
+            buflen = sysconf(_SC_GETGR_R_SIZE_MAX);
+            if (buflen == -1) {
+                buflen = 2048;
+            }
+        } else {
+            // enlarge buffer
+            buflen += 2048;
+        }
+
+        // allocate some buffer
+        mem = realloc(buf, buflen);
+        if (!mem) {
+            WRAPERR(&err, SR_ERR_NO_MEMORY, "realloc()", "")
+            goto cleanup;
+        }
+        buf = mem;
+
+        // GID -> group
+        r = getgrgid_r(gid, &grp, buf, buflen, &grp_p);
+    }
+    while (r == ERANGE);
+
+    if (r) {
+        WRAPERR(&err, SR_ERR_INTERNAL, "Retrieving GID grp entry", strerror(r))
+        goto cleanup;
+    }
+    else if (!grp_p) {
+        WRAPERR(&err, SR_ERR_NOT_FOUND, "Retrieving GID grp entry (No such GID)", "")
+        goto cleanup;
+    }
+
+    // assign group
+    *group = strdup(grp.gr_name);
+    if (!*group) {
+        WRAPERR(&err, SR_ERR_NO_MEMORY, "strdup()", strerror(errno))
+        goto cleanup;
+    }
+
+cleanup:
+    free(buf);
+    return err;
+}
+
+/* standard and no_predicate should be freed, predicate should not */
+static sr_error_info_t *
+srpds_get_predicate(const struct lyd_node *node, char **predicate, char **standard, char **no_predicate)
```

Comment:
If `predicate` shoudl not be freed, it should be `const char *`.

Reply:
done